resources/firmware.rb

```diff
@@ -0,0 +1,90 @@
+# (c) Copyright 2016 Hewlett Packard Enterprise Development LP
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+OneviewCookbook::ResourceBaseProperties.load(self)
+
+property :spp_name, String
+property :spp_file, String
+property :hotfixes_names, Array, default: []
+property :hotfixes_files, Array, default: []
+
+default_action :add
+
+action_class do
+  include OneviewCookbook::Helper
+  include OneviewCookbook::ResourceBase
+
+  def load_firmware(filename = nil)
+    load_sdk
+    c = build_client(client)
+    filename ||= name
+    filename = ::File.basename(filename, ::File.extname(filename)).tr('.', '_')
+    firmware_list = OneviewSDK::FirmwareDriver.find_by(c, {})
+    firmware_list.find { |fw| filename == fw['resourceId'] }
+  end
+end
+
+action :add do
+  item = load_firmware
+
+  if item.nil?
+    c = build_client(client)
+    converge_by "Add #{resource_name} '#{name}'" do
+      OneviewSDK::FirmwareBundle.add(c, name)
+    end
+  else
+    Chef::Log.info "#{resource_name} '#{name}' is already up to date."
+  end
+end
+
+action :create_custom_spp do
+  raise "Unspecified property: 'spp_name' or 'spp_file'. Please set it before attempting this action." unless spp_name || spp_file
+  if hotfixes_names.empty? && hotfixes_files.empty?
+    raise "Unspecified property: 'hotfixes_names' or 'hotfixes_files'. Please set it before attempting this action."
```

Comment:
You're checking if they are **empty** here, but the exception raised says they are **unspecified** (nil)